numba/npyufunc/array_exprs.py

```diff
@@ -26,63 +30,79 @@ def __init__(self, pipeline, *args, **kws):
             special_ops['arrayexpr'] = _lower_array_expr
 
     def match(self, interp, block, typemap, calltypes):
-        '''Using typing and a basic block, search the basic block for array
-        expressions.  Returns True when one or more matches were
-        found, False otherwise.
-        '''
+        """
+        Using typing and a basic block, search the basic block for array
+        expressions.
+        Return True when one or more matches were found, False otherwise.
+        """
         matches = []
-        # We can trivially reject everything if there are fewer than 2
-        # calls in the type results since we'll only rewrite when
-        # there are two or more calls.
-        if len(calltypes) > 1:
+        # We can trivially reject everything if there are no
+        # calls in the type results.
+        if len(calltypes) > 0:
             self.crnt_block = block
             self.typemap = typemap
             self.matches = matches
-            array_assigns = {}
-            self.array_assigns = array_assigns
-            const_assigns = {}
-            self.const_assigns = const_assigns
+            self.array_assigns = {}
```

Comment:
With this patch, we don't need both `array_assigns` and `matches`.  Every `target_name` must be in `array_assigns`.  The only purpose for `matches` is to provide the proper ordering.  Therefore, we can remove `matches` and use `OrderedDict` for `array_assigns`.